### PR TITLE
Network: Add support for parsing multipath routing

### PR DESCRIPTION
**Pull request by anitazha**
Signed-off-by: Anita Zhang <the.anitazha@gmail.com>

## Description
We have some hosts using multipath routing and the network plugin is failing to set any interface information for the interfaces with such multipath configuration. This PR updates `check_routing_table()` in the network plugin to support multipath routing.

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Chore (non-breaking change that does not add functionality or fix an issue)

## Checklist:

- [x] I have read the **CONTRIBUTING** document.
- [x] I have run the pre-merge tests locally and they pass.
- [x] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.
- [x] All commits have been signed-off for [the Developer Certificate of Origin](https://github.com/chef/chef/blob/master/CONTRIBUTING.md#developer-certification-of-origin-dco).

### CODE REVIEW

File: lib/ohai/plugins/linux/network.rb [hunk 1]
```diff
@@ -81,8 +81,16 @@ def check_routing_table(family, iface, default_route_table)
       logger.trace("Plugin Network: Parsing #{line}")
       if /\\/.match?(line)
         parts = line.split("\\")
-        route_dest = parts.shift.strip
+        dest_parts = parts.shift.strip.split("\s", 2)
+        route_dest = dest_parts[0]
         route_endings = parts
```
Comment (jaymzh) [inline lib/ohai/plugins/linux/network.rb:86]: This is pretty confusing way to do this in ruby.

The hidden shift changes parts, and then you assign a pointer of parts to route_endings (everything in ruby is a pointer).

It totally works, but I'd much rather something like this, which I find easier to follow:

```ruby
# If we have multiroute, then the first part will be a normal looking route:
#   default proto ra metric 1024 <other options>
# and then each successive part after that is hops that won't have those options.
# so the first thing we do, is grab that first part, and split it in to the route dest ("default")
# in the above, and then all the options for that route. 
parts = line.split('\\')
route_dest, dest_opts = parts.first.split(nil, 2)
# then all the route endings, generally just hops:
route_endings = parts[1..]
if dest_parts.count > 1
  # your comment here, it's excellent.
  route_endings.map! { |e| e.include?("nexthop") ? "#{dest_opts} #{e}" : e }
end
```

File: spec/unit/plugins/linux/network_spec.rb [hunk 2]
```diff
@@ -1219,6 +1219,38 @@
...
+          expect(plugin["network"]).not_to be_nil
+        end
+
+        it "sets default ipv6 interface and gateway" do
+          expect(plugin["network"]["default_inet6_interface"]).to eq("eth0")
+          expect(plugin["network"]["default_inet6_gateway"]).to eq("fe80::c018:50ff:fe41:ca33")
+        end
+
+        it "sets the routes properly" do
+          expect(plugin["network"]["interfaces"]["eth0"]["routes"]).to eq([Mash.new( destination: "default", family: "inet6", proto: "ra", "via": "fe80::c018:50ff:fe41:ca33", "metric": "1024" )])
```
Comment (jaymzh) [inline spec/unit/plugins/linux/network_spec.rb:1247]: be consistent on your hash syntax
```suggestion
          expect(plugin["network"]["interfaces"]["eth0"]["routes"]).to eq([Mash.new( destination: "default", family: "inet6", proto: "ra", via: "fe80::c018:50ff:fe41:ca33", metric: "1024" )])
```

File: spec/unit/plugins/linux/network_spec.rb [hunk 3]
```diff
@@ -1219,6 +1219,38 @@
...
+          expect(plugin["network"]["interfaces"]["eth0.11"]["routes"]).to eq([Mash.new( destination: "default", family: "inet6", proto: "ra", "via": "fe80::c018:50ff:fe41:ca33", "metric": "1024" )])
+          expect(plugin["network"]["interfaces"]["eth13"]["routes"]).to eq([Mash.new( destination: "default", family: "inet6", proto: "ra", "via": "fe80::c018:50ff:fe3f:9184", "metric": "1050" )])
+          expect(plugin["network"]["interfaces"]["eth3"]["routes"]).to eq([Mash.new( destination: "default", family: "inet6", proto: "ra", "via": "fe80::c018:50ff:fe3f:9184", "metric": "1050" )])
```
Comment (jaymzh) [inline spec/unit/plugins/linux/network_spec.rb:1250]: same

File: lib/ohai/plugins/linux/network.rb [hunk 4]
```diff
@@ -81,8 +81,16 @@ def check_routing_table(family, iface, default_route_table)
       logger.trace("Plugin Network: Parsing #{line}")
       if /\\/.match?(line)
         parts = line.split("\\")
-        route_dest = parts.shift.strip
+        dest_parts = parts.shift.strip.split("\s", 2)
```
Comment (jaymzh) [inline lib/ohai/plugins/linux/network.rb:84]: don't specify `\s`, just use the default splitting which is smart whitespace split.

**Review by jaymzh — CHANGES_REQUESTED** [review]

**Comment by anitazha** [comment]
Thanks Phil! Updated with your suggestions

File: spec/unit/plugins/linux/network_spec.rb [hunk 5]
```diff
@@ -1219,6 +1219,38 @@
         end
       end
 
+      describe "when there're multipath routes" do
+        let(:linux_ip_route) { "" }
```
Comment (jaymzh) [inline spec/unit/plugins/linux/network_spec.rb:1223]: just to make sure people don't break this in the future, can we add a v4 test, please?

**Review by jaymzh — COMMENTED** [review]
Code looks good, sorry I missed the lack of v4-test below, but I suspect that's gonna be the most-commonly-used path in the code, so lets just have a test for it as well?

**Comment by anitazha** [comment]
There was an existing test case for multipath IPv4 but without the extra options on the first line (line 165 of network_spec.rb). The IPv4 test case I added includes options.

**Review by jaymzh — APPROVED** [review]

**Comment by jaymzh** [comment]
Looks great. I'll merge when tests pass. I'm out of town this weekend so feel free to poke me next week if I haven't merged.

**Comment by tas50** [comment]
@anitazha is this something we'd want to backport to Ohai 16 and 17?

**Comment by anitazha** [comment]
@tas50 It could be useful to have since any default IPv6 routes with options using multipath just break without it. But I think multipath routing isn't super widely used so most people won't run into it.

**Pull request merged**